File: src/interpreter.py
# get input
program = input()

# Initialize the memory tape with one cell and set the cell index to 0
tape = [0]
cell_index = 0

# Initialize user input, loop table, and loop stack
user_input = []
loop_table = {}
loop_stack = []

# Build the loop_table to map opening '[' and closing ']' brackets
for ip, instruction in enumerate(program):
    if instruction == "[":
        loop_stack.append(ip)
    elif instruction == "]":
        loop_beginning_index = loop_stack.pop()
        loop_table[loop_beginning_index] = ip
        loop_table[ip] = loop_beginning_index

# Initialize the instruction pointer (ip)
ip = 0

# Execute instructions
while ip < len(program):
    instruction = program[ip]

    if instruction == "+":
        # Increment the value at the current memory cell
        tape[cell_index] += 1
        if tape[cell_index] == 256:
            tape[cell_index] = 0

    elif instruction == "-":
        # Decrement the value at the current memory cell
        tape[cell_index] -= 1
        if tape[cell_index] == -1:
            tape[cell_index] = 255

    elif instruction == "<":
        # Move the memory pointer to the left
        cell_index -= 1

    elif instruction == ">":
        # Move the memory pointer to the right
        cell_index += 1
        if cell_index == len(tape):
            tape.append(0)  # Add another cell if we move beyond the end of the tape

    elif instruction == ".":
        # Output the ASCII value of the current cell as a character
        print(chr(tape[cell_index]), end="")

    elif instruction == ",":
        # Input a character and store its ASCII value in the current memory cell
        if user_input == []:
            user_input = list(input() + "\n")
        tape[cell_index] = ord(user_input.pop(0))

    elif instruction == "[":
        if not tape[cell_index]:
            # Jump past the closing ']' if the current memory cell is 0
            ip = loop_table[ip]

    elif instruction == "]":
        if tape[cell_index]:
            # Jump back to the matching opening '[' if the current memory cell is not 0
            ip = loop_table[ip]

    # Move to the next instruction
    ip += 1

File: src/C/interpreter.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TAPE_SIZE 30000
// this follows the more traditional tape rather than the python interpreter

void execute_brainfuck_program(const char* program) {
    unsigned char tape[TAPE_SIZE] = {0};
    size_t cell_index = 0;
    size_t ip = 0;

    // Build the loop_table to map opening '[' and closing ']' brackets
    size_t loop_table[TAPE_SIZE] = {0};
    size_t loop_stack[TAPE_SIZE] = {0};
    size_t loop_stack_index = 0;

    for (size_t i = 0; i < strlen(program); i++) {
        char instruction = program[i];
        if (instruction == '[') {
            loop_stack[loop_stack_index++] = i;
        } else if (instruction == ']') {
            if (loop_stack_index > 0) {
                size_t open_bracket_index = loop_stack[--loop_stack_index];
                loop_table[i] = open_bracket_index;
                loop_table[open_bracket_index] = i;
            }
        }
    }

    while (ip < strlen(program)) {
        char instruction = program[ip];

        switch (instruction) {
            case '+':
                // Increment the value at the current memory cell
                tape[cell_index]++;
                break;
            case '-':
                // Decrement the value at the current memory cell
                tape[cell_index]--;
                break;
            case '<':
                // Move the memory pointer to the left
                if (cell_index > 0) {
                    cell_index--;
                }
                break;
            case '>':
                // Move the memory pointer to the right
                cell_index++;
                if (cell_index >= TAPE_SIZE) {
                    printf("Tape overflow!\n");
                    exit(1);
                }
                break;
            case '.':
                // Output the ASCII value of the current cell as a character
                putchar(tape[cell_index]);
                break;
            case ',':
                // Input a character and store its ASCII value in the current memory cell
                tape[cell_index] = getchar();
                break;
            case '[':
                // Jump past the closing ']' if the current memory cell is 0
                if (tape[cell_index] == 0) {
                    ip = loop_table[ip];
                }
                break;
            case ']':
                // Jump back to the matching opening '[' if the current memory cell is not 0
                if (tape[cell_index] != 0) {
                    ip = loop_table[ip];
                }
                break;
            default:
                // Ignore other characters
                // Good enough for now but will need to add either comments or errors
                break;
        }

        ip++;
    }
}

int main() {
    char program[10000]; // Adjust the size as needed
    printf("Enter Brainfuck code: ");
    fgets(program, sizeof(program), stdin);

    execute_brainfuck_program(program);

    return 0;
}